### DISCUSSION TITLE
Leaving expressions with undefined variables

### DISCUSSION THREAD

**Discussion by vvladic** [pallets/jinja#1908] [Q&A]
Hello community,

First of all, thank you for developing and maintaining jinja2.

I need your help :-)

I have a template: 
`some text {{ a }} more text {{ b + 1 }} text text {{ b + c.d }}`

I want to achieve the following behavior:
Render the expression if it can be resolved, otherwise just leave it there, so I can render it later, in the next stage.
For example, with `a=3`, and `b` left undefined, I need the following result:
`some text 3 more text {{ b + 1 }} text text {{ b + c.d }}`

Is there any straightforward way to achieve this behavior?
I tried subclassing `class Undefined`, but the problem is, I only have access to the missing variable name, and not the whole expression.

I tried doing the steps in `Template.render()` manually (parse, generate, compile), to understand what is actually happening behind the scenes. Before trying to subclass `class CodeGenerator`, I wanted to ask if there was a better way to achieve that.

Thank you.

**Comment by davidism** [comment]
You could use [`find_undeclared_variables`](https://jinja.palletsprojects.com/en/3.1.x/api/#jinja2.meta.find_undeclared_variables) and skip rendering until later if you don't have all the values needed yet.

**Reply by vvladic** [reply]
Hi David, 

thank you for your answer.

The problem with this is that I have multiple (two or more) stages where I need to partially render my template.
At the first stage, only variable `a` is available, at the second stage only variable `b` is available, while variable `a` is not available any more, and so on.

I could do something like this in each stage:
```
env = Environment(undefined=CustomUndefined)
env.globals.update(self.vars)
...
expressions = set(re.findall("{{.*?}}", text))
undeclared = find_undeclared_variables(env.parse(text))
    for expression in expressions:
        if any(var in expression for var in undeclared):
            text = text.replace(expression, f"{{% raw %}}{expression}{{% endraw %}}")

template = env.from_string(text)
return template.render()
```
but then I would have to put my variables in the `env.globals` and create new Environment each time.
Is there a way to access the original expression string from my undefined object or from inside of the generated root function?
As I need to compile the template for each stage anyway, then I could also only generate, maybe modify and execute the root function without compiling it to bytecode?
The problem is, how can I get my whole expression string that corresponds with each undefined object in that expression.
